### ISSUE TITLE
Open-MAGVIT2 achieves SOTA??? I just cannot reproduce the numbers reported in the main page.

### ISSUE THREAD

**Issue by juncongmoo** [TencentARC/Open-MAGVIT2#13]
I tried with the evaluation.py script and the pre-trained checkpoints provided in the github repo's main page. However, I could not get the same rFID,and PSNR number displayed in the page(the ultilization seems almost 100% though). The number I got is 10+ for rFID, and 6 for PSNR. 

In the main page, the rFID should be 1.53 and 0.39 and PSNR should be 21 and 25. My number is 10+, and 6+, which are so far away from it....They are even worse than vanilla VQGAN. I used official scripts and checkpoints. Does anybody get the same number? Or did I miss anything?

**Comment by shinshiner** [comment]
same question, with 100,000 imagenet test images, have you solve it?

**Comment by RobertLuo1** [comment]
Hi, we rerun the evaluation code with the provided setting (128×128 imagenet tokenizer) on the 128 × 128 imagenet 5k validation set. The result is consistent with the main page. @shinshiner We follow the previous prevalent setting that perform evaluation using 5k validation set. We have not conducted any evaluation on the test set. @juncongmoo I wonder if there is something wrong when specifying the model path corresponding to the training image resolution for evaluation.

**Comment by shinshiner** [comment]
> Hi, we rerun the evaluation code with the provided setting (128×128 imagenet tokenizer) on the 128 × 128 imagenet 50k validation set. The result is consistent with the main page. @shinshiner We follow the previous prevalent setting that perform evaluation using 50k validation set. We have not conducted any evaluation on the test set. @juncongmoo I wonder if there is something wrong when specifying the model path corresponding to the training image resolution for evaluation. Furthermore, we show the output log of the result. ![image](https://private-user-images.githubusercontent.com/72377255/344689468-85daeb0c-765a-4855-9467-8b7395bc0e98.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwMDk3NzIsIm5iZiI6MTcyMDAwOTQ3MiwicGF0aCI6Ii83MjM3NzI1NS8zNDQ2ODk0NjgtODVkYWViMGMtNzY1YS00ODU1LTk0NjctOGI3Mzk1YmMwZTk4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzAzVDEyMjQzMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0Y2I1YzhhYzNmZTI5YWMxMmMxYTFjMjQxNmM3MjBmNDM4NzdlZjQwNTYyYjcwN2Y2OWRiNzhjNGVkOWNmYWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.zv8nJdyhnKEsGaq_6pBjZEE9igfQsuhirTXF1dqMavw)

Hello, I fix the settings and evaluate my trained model in validation set, but I found both SSIM and PSNR are higher than yours (0.82 & 25.90), however, the FID and LPIPS is extremely worser (20 & 0.23). I'm not sure what it stands for and what problems inside.

**Comment by RobertLuo1** [comment]
Hello, Can you provide more details of your trained model? Although PSNR and SSIM have no absolute relationship with the rFID and LPIPS, they are somehow relevant. It is hard to target the problems based on the information you provide. With such high PSNR and SSIM, I think you can visualize the reconstructed images to see if they are too smooth.

**Comment by shinshiner** [comment]
> Hello, Can you provide more details of your trained model? Although PSNR and SSIM have no absolute relationship with the rFID and LPIPS, they are somehow relevant. It is hard to target the problems based on the information you provide. With such high PSNR and SSIM, I think you can visualize the reconstructed images to see if they are too smooth.

The reconstruction results are indeed too smooth (as shown in attach files with 128*128, gt in left), and crashed in some details, especially face part), what a strange...
Maybe I need to tune some of your hyper parameters if I use different number of gpus or batch size?
![image_recons_test_1300](https://github.com/TencentARC/Open-MAGVIT2/assets/24225683/9275d211-505a-4ab6-82f1-97445b0357b1)
![image_recons_test_10140](https://github.com/TencentARC/Open-MAGVIT2/assets/24225683/1a5dd5c4-05c0-422b-8564-b298c9ad16de)
![image_recons_test_11970](https://github.com/TencentARC/Open-MAGVIT2/assets/24225683/073deaa0-510d-4dd1-8628-dcb436fa08a4)

**Comment by RobertLuo1** [comment]
The reconstruction of face part is relatively hard because of the high compression rate of the tokenizer and the training data. I think you need to monitor the GAN training when using the different GPUs and batchsizes.

**Comment by shinshiner** [comment]
> The reconstruction of face part is relatively hard because of the high compression rate of the tokenizer and the training data. I think you need to monitor the GAN training when using the different GPUs and batchsizes.

Not only faces, but almost all types of details are lost. I visualized the GAN training process, and found the discriminator loss is decreasing while generator loss is oppsite. And the pixel-reconstruction loss and perceputal loss is increasing too! 

It looks like the discriminator is too strong, and in such way, generator gets no information to perform better from discriminator. But I cannot understand why the reconstruction loss is increasing, even no grad from adversial loss, it can still be optimized by reconstruction and perceptual loss, and the two losses should be decreasing in training process. This is what confused me most. Do you have any experience to finetune GAN training of such case? @RobertLuo1 

I've tried to slow down the training of discriminator (by learning rate, or optimization intervals), but it only delayed the crashed training, cannot prevent it :(

**Comment by henrywoo** [comment]
The reconstruction results are really bad. I used the `inference.py` file from the repo and checkpoint from [here](https://huggingface.co/TencentARC/Open-MAGVIT2/blob/main/imagenet_128_B.ckpt).

![image](https://github.com/TencentARC/Open-MAGVIT2/assets/458995/9561756a-6ff6-4f7c-8845-8796f8c07258)

Worse than the main page's results. Even worse than above comment's ones. Why?

**Comment by henrywoo** [comment]
I tested the pretrained with the imagenet-1k dataset from [here](https://huggingface.co/datasets/ILSVRC/imagenet-1k).
There are 50K validation images. Can you please confirm the checkpoints released to public is the same one that you got the 1.56 RFID from? Thanks.

```
$md5sum imagenet_128_B.ckpt
1c71cd9e06742ae6c30f6f69386d8776  mbin/imagenet_128_B.ckpt
$python evaluation.py --config_model configs/imagenet_lfqgan_128_B.yaml --ckpt_path imagenet_128_B.ckpt
```

Result is:

![image](https://github.com/TencentARC/Open-MAGVIT2/assets/458995/1c06747e-fc17-4cd2-91b9-ea8b2406ce39)

**Comment by RobertLuo1** [comment]
Hi, @henrywoo, we directly download the checkpoints from [here](https://huggingface.co/TencentARC/Open-MAGVIT2/blob/main/imagenet_128_B.ckpt) and re-evaluate it again. The results are still the same as I mentioned before. For further potential problem reproduction, we use the checkpoints trained with $256 \times 256$ imagenet for inference on the case you provide. Since we can not get the original image, we use the screenshot as the input image, which is quite vague and of low quality. However, the reconstructed image is still normal and the phenomenon does not happen. (See the figure below, the former image is the original image and the latter one is reconstructed from the model.)
![0](https://github.com/TencentARC/Open-MAGVIT2/assets/72377255/1575c4a7-a2e6-4eab-bc18-6a42a8ce552b)
![0](https://github.com/TencentARC/Open-MAGVIT2/assets/72377255/8fe7baa6-35cf-4558-b630-5c2a49b3816d)
When I dive deeper into evaluation.py, I find that it may not be quite convenient for the checkpoint path configuration. I have already updated it.

**Comment by RobertLuo1** [comment]
Hi, @henrywoo Thanks for your codes! I have looked through your code and found something different with us.  
1) https://github.com/henrywoo/Open-MAGVIT2/blob/f848114e4b56e187de98b4a9c56b8851e08d87c1/inet1k.py#L37 Actually, when we train our model, we use mean=[0.5, 0.5, 0.5] and std=[0.5, 0.5, 0.5] instead, which is well utilized in tokenizer training, e.g., VQGAN, VAR, LlamaGEN.
2) We use Resize and then CenterCrop for the Data Preprocessing when evaluation, which is also adopted from previous works.
I think the CenterCrop will not make a huge impact on the rFID. Instead, the wrong Normalization may be the main factor.

**Comment by RobertLuo1** [comment]
@shinshiner In our training version, the D and G loss will increase and decrease alternately. The whole tendency of G loss is downward while the D loss is the opposite. The Reconstruction and Perceptual Loss is always decreasing. 
I think the reason why Recon and Lpips Loss increase stems from the unstable GAN training as you mentioned that the discriminator is too strong. However, we have not encountered that case.

**Comment by RobertLuo1** [comment]
@henrywoo Henry, we follow our assumption and use the wrong data preprocessing for evaluation. We are able to get the results (high rFID) relatively close to you. I think if you change back to the correct pre-processing, the result will be consistent with the main page. @juncongmoo, It seems that you have encountered the same issues with Henry. Could you please check your code again?
![image](https://github.com/TencentARC/Open-MAGVIT2/assets/72377255/0d761c30-dc84-4d13-8bcc-d0cc03a8c925)

**Comment by shinshiner** [comment]
> @shinshiner In our training version, the D and G loss will increase and decrease alternately. The whole tendency of G loss is downward while the D loss is the opposite. The Reconstruction and Perceptual Loss is always decreasing. I think the reason why Recon and Lpips Loss increase stems from the unstable GAN training as you mentioned that the discriminator is too strong. However, we have not encountered that case.

could you share your loss curve for reference? I use less gpus and thus smaller batch size, is there any clue to finetune the hyper parameters?

**Comment by henrywoo** [comment]
> Hi, @henrywoo Thanks for your codes! I have looked through your code and found something different with us.
> 
> 1. https://github.com/henrywoo/Open-MAGVIT2/blob/f848114e4b56e187de98b4a9c56b8851e08d87c1/inet1k.py#L37 Actually, when we train our model, we use mean=[0.5, 0.5, 0.5] and std=[0.5, 0.5, 0.5] instead, which is well utilized in tokenizer training, e.g., VQGAN, VAR, LlamaGEN.
> 2. We use Resize and then CenterCrop for the Data Preprocessing when evaluation, which is also adopted from previous works.
>    I think the CenterCrop will not make a huge impact on the rFID. Instead, the wrong Normalization may be the main factor.

Thanks a lot @RobertLuo1 It works with all mean and std as 0.5 like what you said. I remember we should use ImageNet-specific normalization for imagenet dataset training, not sure why in this case, 0.5 is better. But anyway, it works.
